### PR TITLE
Make opponent's hand to appear randomized to the player when needed

### PR DESCRIPTION
**Pull request by starg09**
Fixes potential exploits where a player may use a hand-seeing power, then one that is meant to be random or hidden from their view. Without randomization they can predict which card to discard specifically, breaking many decks in the process.

Work on this must be done for legacy sets ASAP in my opinion. But this should solve a good amount of those interaction, as long as at least one groovy-coded card is involved.

### CODE REVIEW

File: src/tcgwars/logic/impl/gen7/TeamUp.groovy [hunk 1]
```diff
@@ -1423,10 +1423,13 @@ public enum TeamUp implements LogicCardInfo {
...
             energyCost P,P
+            attackRequirement {
+              assert opp.hand : "Your opponent has no cards in their hand"
+            }
             onAttack{
-              if(opp.hand){
-                opp.hand.showToMe("Your opponent's hand")
-                damage 50*opp.hand.filterByType(TRAINER).size()
+              if (opp.hand){
+                trainersInHand = opp.hand.shuffledCopy().showToMe("Your opponent's hand").filterByType(TRAINER)
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen7/TeamUp.groovy:1431]: `trainersInHand` isn't defined here yet

File: src/tcgwars/logic/impl/gen7/TeamUp.groovy [hunk 2]
```diff
@@ -3076,11 +3082,11 @@ public enum TeamUp implements LogicCardInfo {
...
               afterDamage{
                 if(opp.hand.size() >= 4){
-                  if(opp.hand.size() == 4){
-                    opp.hand.showToMe("Your opponent's hand")
-                  }
-                  else{
-                    opp.hand.select(count:opp.hand.size() - 4,"Choose the cards to discard").discard()
+                  def randomOppHand = opp.hand.shuffledCopy()
+                  if (opp.hand.size() == 4){
+                    randomOppHand.shuffledCopy().showToMe("Your opponent's hand.")
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen7/TeamUp.groovy:3087]: is the second `shuffledCopy()` operation needed here?

File: src/tcgwars/logic/util/CardList.groovy [hunk 3]
```diff
@@ -61,6 +61,11 @@ public class CardList extends ArrayList<Card> {
   public CardList copy(){
     return new CardList(this);
   }
+  public CardList shuffledCopy(){
```
Comment (luongthomas) [inline src/tcgwars/logic/util/CardList.groovy:64]: This is an extra shuffledCopy(), there is another function already defined in line 70

File: src/tcgwars/logic/impl/gen7/TeamUp.groovy (hunk 2 above)
Comment (starg09) [reply src/tcgwars/logic/impl/gen7/TeamUp.groovy:3087]: Nope, leftover from before starting to use a variable. Nice catch 👍

File: src/tcgwars/logic/util/CardList.groovy (hunk 3 above)
Comment (starg09) [reply src/tcgwars/logic/util/CardList.groovy:64]: Welp, guess that's one thing to lookout when rebase-merging these PRs (tested a card in master with the method)

File: src/tcgwars/logic/impl/gen7/ForbiddenLight.groovy [hunk 4]
```diff
@@ -2158,14 +2158,18 @@ public enum ForbiddenLight implements LogicCardInfo {
...
             }
             onAttack {
-              if(opp.hand.hasType(SUPPORTER)){
-                def card=opp.hand.select("Opponent's hand. Select a supporter.", cardTypeFilter(SUPPORTER)).first()
-                discard card
-                bg.deterministicCurrentThreadPlayerType=self.owner
-                bg.em().run(new PlayTrainer(card))
-                bg.clearDeterministicCurrentThreadPlayerType()
+              if (opp.hand) {
+              def randomOppHand = opp.hand.shuffledCopy()
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen7/ForbiddenLight.groovy:2162]: Could you indent this `if` block over by one?  Also it looks like it's missing a closing bracket
